diff --git a/openbsd-compat/openssl-compat.c b/openbsd-compat/openssl-compat.c
index 1486507..e196547 100644
--- a/openbsd-compat/openssl-compat.c
+++ b/openbsd-compat/openssl-compat.c
@@ -121,4 +121,227 @@ EVP_DigestVerify(EVP_MD_CTX *ctx, const unsigned char *sigret, size_t siglen,
 }
 #endif
 
+#if defined(LIBRESSL_VERSION_NUMBER) || defined(OPENSSL_IS_BORINGSSL)
+//XXX(ossl3): proper feature test
+int PEM_write_bio_PrivateKey_traditional(BIO *bp, EVP_PKEY *x,
+    const EVP_CIPHER *enc, unsigned char *kstr, int klen, pem_password_cb *cb,
+    void *u)
+{
+	if (EVP_PKEY_base_id(x) == EVP_PKEY_DSA) {
+		return PEM_write_bio_DSAPrivateKey(bp, EVP_PKEY_get0_DSA(x),
+		    enc, kstr, klen, cb, u);
+	} else if (EVP_PKEY_base_id(x) == EVP_PKEY_RSA) {
+		return PEM_write_bio_RSAPrivateKey(bp, EVP_PKEY_get0_RSA(x),
+		    enc, kstr, klen, cb, u);
+	}
+#ifdef OPENSSL_HAS_ECC
+	else if (EVP_PKEY_base_id(x) == EVP_PKEY_EC) {
+		return PEM_write_bio_ECPrivateKey(bp, EVP_PKEY_get0_EC_KEY(x),
+		    enc, kstr, klen, cb, u);
+	}
+#endif
+	return 0;
+}
+#endif
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+//XXX(ossl3): separate method tests
+static int EVP_PKEY_is_a(const EVP_PKEY *pkey, const char *name)
+{
+	int rc = 0;
+
+	if (strcmp(name, "DSA") == 0 &&
+			EVP_PKEY_get0_DSA((EVP_PKEY *) pkey) != NULL) {
+		rc = 1;
+	} else if (strcmp(name, "RSA") == 0 &&
+			EVP_PKEY_get0_RSA((EVP_PKEY *) pkey) != NULL) {
+		rc = 1;
+	}
+
+	return rc;
+}
+
+static int EVP_PKEY_get_bn_param(const EVP_PKEY *pkey,
+				 const char *key_name,
+				 BIGNUM **bn)
+{
+	DSA *dsa = EVP_PKEY_is_a(pkey, "DSA") ?
+	    EVP_PKEY_get0_DSA((EVP_PKEY *) pkey) : NULL;
+	RSA *rsa = EVP_PKEY_is_a(pkey, "RSA") ?
+	    EVP_PKEY_get0_RSA((EVP_PKEY *) pkey) : NULL;
+	const BIGNUM *val = NULL;
+	int rc = 0;
+
+	if (dsa != NULL) {
+		if (strcmp(key_name, OSSL_PKEY_PARAM_PUB_KEY) == 0) {
+			DSA_get0_key(dsa, &val, NULL);
+		} else if (strcmp(key_name, OSSL_PKEY_PARAM_PRIV_KEY) == 0) {
+			DSA_get0_key(dsa, NULL, &val);
+		} else if (strcmp(key_name, OSSL_PKEY_PARAM_FFC_P) == 0) {
+			DSA_get0_pqg(dsa, &val, NULL, NULL);
+		} else if (strcmp(key_name, OSSL_PKEY_PARAM_FFC_Q) == 0) {
+			DSA_get0_pqg(dsa, NULL, &val, NULL);
+		} else if (strcmp(key_name, OSSL_PKEY_PARAM_FFC_G) == 0) {
+			DSA_get0_pqg(dsa, NULL, NULL, &val);
+		}
+	} else if (rsa != NULL) {
+		if (strcmp(key_name, OSSL_PKEY_PARAM_RSA_N) == 0) {
+			RSA_get0_key(rsa, &val, NULL, NULL);
+		} else if (strcmp(key_name, OSSL_PKEY_PARAM_RSA_E) == 0) {
+			RSA_get0_key(rsa, NULL, &val, NULL);
+		} else if (strcmp(key_name, OSSL_PKEY_PARAM_RSA_D) == 0) {
+			RSA_get0_key(rsa, NULL, NULL, &val);
+		} else if (strcmp(key_name,
+			   OSSL_PKEY_PARAM_RSA_COEFFICIENT1) == 0) {
+			RSA_get0_crt_params(rsa, NULL, NULL, &val);
+		} else if (strcmp(key_name,
+			   OSSL_PKEY_PARAM_RSA_FACTOR1) == 0) {
+			RSA_get0_factors(rsa, &val, NULL);
+		} else if (strcmp(key_name,
+			   OSSL_PKEY_PARAM_RSA_FACTOR2) == 0) {
+			RSA_get0_factors(rsa, NULL, &val);
+		}
+	}
+
+	if (val != NULL) {
+		*bn = BN_dup(val);
+		rc = 1;
+	}
+
+	return rc;
+}
+
+int EVP_PKEY_print_public_fp(FILE *fp, const EVP_PKEY *pkey, int indent,
+    ASN1_PCTX *pctx)
+{
+	int ret = 0;
+	BIO *out = BIO_new_fd(fileno(fd), BIO_NOCLOSE);
+	if (out != NULL) {
+		ret = EVP_PKEY_print_public(out, pkey, indent, pctx);
+		BIO_free(out);
+	}
+	return ret;
+}
+
+int EVP_PKEY_print_private_fp(FILE *fp, const EVP_PKEY *pkey, int indent,
+    ASN1_PCTX *pctx)
+{
+	int ret = 0;
+	BIO *out = BIO_new_fd(fileno(fd), BIO_NOCLOSE);
+	if (out != NULL) {
+		ret = EVP_PKEY_print_private(out, pkey, indent, pctx);
+		BIO_free(out);
+	}
+	return ret;
+}
+
+int EVP_PKEY_print_params_fp(FILE *fp, const EVP_PKEY *pkey, int indent,
+    ASN1_PCTX *pctx)
+{
+	int ret = 0;
+	BIO *out = BIO_new_fd(fileno(fd), BIO_NOCLOSE);
+	if (out != NULL) {
+		ret = EVP_PKEY_print_params(out, pkey, indent, pctx);
+		BIO_free(out);
+	}
+	return ret;
+}
+
+size_t EVP_PKEY_get1_encoded_public_key(EVP_PKEY *pkey, unsigned char **ppub)
+{
+	if (EVP_PKEY_base_id(a) == EVP_PKEY_EC)
+		return EVP_PKEY_get1_tls_encodedpoint(pkey, ppub);
+	return 0;
+}
+
+int EVP_PKEY_set1_encoded_public_key(EVP_PKEY *pkey,
+    const unsigned char *pub, size_t publen)
+{
+	if (EVP_PKEY_base_id(a) == EVP_PKEY_EC)
+		return EVP_PKEY_set1_tls_encodedpoint(pkey, pub, publen);
+	if (EVP_PKEY_base_id(a) == EVP_PKEY_DH) {
+		DH *dh = NULL;
+		const BIGNUM *dh_p;
+		BIGNUM *pub_key = NULL;
+		if (pkey == NULL ||
+		    (dh = EVP_PKEY_get0_DH(pkey)) == NULL)
+			return 0;
+		DH_get0_pqg(dh, &dh_p, NULL, NULL);
+		if (dh_p == NULL || BN_num_bytes(dh_p) == 0)
+			return 0;
+		if ((pub_key = BN_bin2bn(pub, (int) publen, NULL)) != NULL) {
+			int result = 0;
+			if (DH_check_pub_key(dh, pub_key, &result) == 1 &&
+			    result == 0 &&
+			    DH_set0_key(dh, pub_key, NULL) == 1) {
+				return 1;
+			}
+			BN_clear_free(pub_key);
+			return 0;
+		}
+		return 0;
+	}
+	return 0;
+}
+
+int i2d_KeyParams(const EVP_PKEY *a, unsigned char **pp)
+{
+	if (EVP_PKEY_base_id(a) == EVP_PKEY_EC)
+		return i2d_ECParameters(EVP_PKEY_get0_EC_KEY(a), pp);
+	if (EVP_PKEY_base_id(a) == EVP_PKEY_DH)
+		return i2d_DHparams(EVP_PKEY_get0_DH(a), pp);
+	ERR_raise(ERR_LIB_ASN1, ASN1_R_UNSUPPORTED_TYPE);
+	return -1;
+}
+
+EVP_PKEY *d2i_KeyParams(int type, EVP_PKEY **a, const unsigned char **pp
+    long length)
+{
+	EVP_PKEY *ret = NULL;
+
+	if ((a == NULL) || (*a == NULL)) {
+		if ((ret = EVP_PKEY_new()) == NULL)
+			return NULL;
+	else
+		ret = *a;
+
+	if (type != EVP_PKEY_id(ret) || EVP_PKEY_set_type(ret, type) == 0)
+		goto err;
+
+	if (type == EVP_PKEY_EC) {
+		EC_KEY *k;
+		if ((k = d2i_ECParameters(NULL, pp, length)) == NULL)
+			goto err;
+		EVP_PKEY_assign_EC_KEY(ret, k);
+	} else if (type == EVP_PKEY_DH) {
+		DH *k;
+		if ((k = d2i_DHparams(NULL, pp, length)) == NULL)
+			goto err;
+		EVP_PKEY_assign_DH(ret, k);
+	} else {
+		ERR_raise(ERR_LIB_ASN1, ASN1_R_UNSUPPORTED_TYPE);
+		goto err;
+	}
+
+	if (a != NULL)
+		*a = ret;
+	return ret;
+err:
+	if ((a == NULL) || *a != ret)
+		EVP_PKEY_free(ret);
+	return NULL;
+}
+
+int EC_GROUP_get_field_type(const EC_GROUP *g)
+{
+	return EC_METHOD_get_field_type(EC_GROUP_method_of(g));
+}
+
+int BN_check_prime(const BIGNUM *p, BN_CTX *ctx, BN_GENCB *cb)
+{
+	int checks = BN_num_bits(p) > 2048 ? 128 : 64;
+	return BN_check_prime_fasttest_ex(p, checks, ctx, 1, cb);
+}
+#endif
+
 #endif /* WITH_OPENSSL */
diff --git a/openbsd-compat/openssl-compat.h b/openbsd-compat/openssl-compat.h
index 2b9780f..f0b1533 100644
--- a/openbsd-compat/openssl-compat.h
+++ b/openbsd-compat/openssl-compat.h
@@ -29,6 +29,13 @@
 #include <openssl/ecdsa.h>
 #endif
 #include <openssl/dh.h>
+#include <openssl/bn.h>
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
+#include <openssl/core_names.h>
+#include <openssl/decoder.h>
+#include <openssl/encoder.h>
+#include <openssl/param_build.h>
+#endif
 
 int ssh_compatible_openssl(long, long);
 void ssh_libcrypto_init(void);
@@ -88,5 +95,47 @@ int EVP_DigestVerify(EVP_MD_CTX *, const unsigned char *, size_t,
     const unsigned char *, size_t);
 #endif
 
+#if defined(LIBRESSL_VERSION_NUMBER) || defined(OPENSSL_IS_BORINGSSL)
+//XXX(ossl3): proper feature test
+int PEM_write_bio_PrivateKey_traditional(BIO *bp, EVP_PKEY *x,
+    const EVP_CIPHER *enc, unsigned char *kstr, int klen, pem_password_cb *cb,
+    void *u);
+#endif
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+//XXX(ossl3): test for each method separately
+#define OSSL_PKEY_PARAM_PUB_KEY "pub"
+#define OSSL_PKEY_PARAM_PRIV_KEY "priv"
+#define OSSL_PKEY_PARAM_FFC_P "p"
+#define OSSL_PKEY_PARAM_FFC_Q "q"
+#define OSSL_PKEY_PARAM_FFC_G "g"
+#define OSSL_PKEY_PARAM_RSA_N "n"
+#define OSSL_PKEY_PARAM_RSA_E "e"
+#define OSSL_PKEY_PARAM_RSA_D "d"
+#define OSSL_PKEY_PARAM_RSA_COEFFICIENT1 "rsa-coefficient1"
+#define OSSL_PKEY_PARAM_RSA_FACTOR1 "rsa-factor1"
+#define OSSL_PKEY_PARAM_RSA_FACTOR2 "rsa-factor2"
+#define EC_POINT_get_affine_coordinates EC_POINT_get_affine_coordinates_GFp
+#define EC_POINT_set_affine_coordinates EC_POINT_set_affine_coordinates_GFp
+#define EVP_PKEY_eq EVP_PKEY_cmp
+int EVP_PKEY_is_a(const EVP_PKEY *pkey, const char *name);
+int EVP_PKEY_get_bn_param(const EVP_PKEY *pkey, const char *key_name,
+    BIGNUM **bn);
+int EVP_PKEY_print_public_fp(FILE *fp, const EVP_PKEY *pkey, int indent,
+    ASN1_PCTX *pctx);
+int EVP_PKEY_print_private_fp(FILE *fp, const EVP_PKEY *pkey, int indent,
+    ASN1_PCTX *pctx);
+int EVP_PKEY_print_params_fp(FILE *fp, const EVP_PKEY *pkey, int indent,
+    ASN1_PCTX *pctx);
+size_t EVP_PKEY_get1_encoded_public_key(EVP_PKEY *pkey, unsigned char **ppub);
+int EVP_PKEY_set1_encoded_public_key(EVP_PKEY *pkey,
+    const unsigned char *pub, size_t publen);
+int i2d_KeyParams(const EVP_PKEY *a, unsigned char **pp);
+EVP_PKEY *d2i_KeyParams(int type, EVP_PKEY **a, const unsigned char **pp
+    long length);
+int EC_GROUP_get_field_type(const EC_GROUP *g);
+int BN_check_prime(const BIGNUM *p, BN_CTX *ctx, BN_GENCB *cb);
+#endif /* OPENSSL_VERSION_NUMBER < 0x30000000L */
+
 #endif /* WITH_OPENSSL */
 #endif /* _OPENSSL_COMPAT_H */
