diff --git a/configure.ac b/configure.ac
index 9053a9a..f4f802a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2893,7 +2893,7 @@ if test "x$openssl" = "xyes" ; then
 			30*)
 				# OpenSSL 3; we use the 1.1x API
 				# https://openssl.org/policies/general/versioning-policy.html
-				CPPFLAGS="$CPPFLAGS -DOPENSSL_API_COMPAT=0x10100000L"
+				CPPFLAGS="$CPPFLAGS -DOPENSSL_API_COMPAT=30000 -DOPENSSL_NO_DEPRECATED"
 				;;
 		        *)
 				AC_MSG_ERROR([Unknown/unsupported OpenSSL version ("$ssl_showver")])
@@ -2956,7 +2956,7 @@ if test "x$openssl" = "xyes" ; then
 	AC_MSG_CHECKING([if programs using OpenSSL functions will link])
 	AC_LINK_IFELSE(
 		[AC_LANG_PROGRAM([[ #include <openssl/err.h> ]],
-		[[ ERR_load_crypto_strings(); ]])],
+		[[ ERR_clear_error(); ]])],
 		[
 			AC_MSG_RESULT([yes])
 		],
@@ -2966,7 +2966,7 @@ if test "x$openssl" = "xyes" ; then
 			AC_MSG_CHECKING([if programs using OpenSSL need -ldl])
 			AC_LINK_IFELSE(
 				[AC_LANG_PROGRAM([[ #include <openssl/err.h> ]],
-				[[ ERR_load_crypto_strings(); ]])],
+				[[ ERR_clear_error(); ]])],
 				[
 					AC_MSG_RESULT([yes])
 					CHANNELLIBS="$CHANNELLIBS -ldl"
@@ -3081,7 +3081,7 @@ if test "x$openssl" = "xyes" ; then
 	#include <openssl/objects.h>
 	#include <openssl/opensslv.h>
 		]], [[
-		EC_KEY *e = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+		EC_GROUP *e = EC_GROUP_new_by_curve_name(NID_X9_62_prime256v1);
 		const EVP_MD *m = EVP_sha256(); /* We need this too */
 		]])],
 		[ AC_MSG_RESULT([yes])
@@ -3099,7 +3099,7 @@ if test "x$openssl" = "xyes" ; then
 	#include <openssl/objects.h>
 	#include <openssl/opensslv.h>
 		]], [[
-		EC_KEY *e = EC_KEY_new_by_curve_name(NID_secp384r1);
+		EC_GROUP *e = EC_GROUP_new_by_curve_name(NID_secp384r1);
 		const EVP_MD *m = EVP_sha384(); /* We need this too */
 		]])],
 		[ AC_MSG_RESULT([yes])
@@ -3117,7 +3117,7 @@ if test "x$openssl" = "xyes" ; then
 	#include <openssl/objects.h>
 	#include <openssl/opensslv.h>
 		]], [[
-		EC_KEY *e = EC_KEY_new_by_curve_name(NID_secp521r1);
+		EC_GROUP *e = EC_GROUP_new_by_curve_name(NID_secp521r1);
 		const EVP_MD *m = EVP_sha512(); /* We need this too */
 		]])],
 		[ AC_MSG_RESULT([yes])
@@ -3132,7 +3132,7 @@ if test "x$openssl" = "xyes" ; then
 	#include <openssl/objects.h>
 	#include <openssl/opensslv.h>
 			]],[[
-			EC_KEY *e = EC_KEY_new_by_curve_name(NID_secp521r1);
+			EC_GROUP *e = EC_GROUP_new_by_curve_name(NID_secp521r1);
 			const EVP_MD *m = EVP_sha512(); /* We need this too */
 			exit(e == NULL || m == NULL);
 			]])],
