Index: os-openssh/Makefile.in
===================================================================
--- os-openssh.orig/Makefile.in
+++ os-openssh/Makefile.in
@@ -116,7 +116,7 @@ LIBSSH_OBJS=${LIBOPENSSH_OBJS} \
 	kexgexc.o kexgexs.o \
 	kexsntrup761x25519.o kexmlkem768x25519.o sntrup761.o kexgen.o \
 	sftp-realpath.o platform-pledge.o platform-tracing.o platform-misc.o \
-	sshbuf-io.o
+	sshbuf-io.o fips.o
 
 SKOBJS=	ssh-sk-client.o
 
Index: os-openssh/cipher.c
===================================================================
--- os-openssh.orig/cipher.c
+++ os-openssh/cipher.c
@@ -48,6 +48,7 @@
 #include "sshbuf.h"
 #include "ssherr.h"
 #include "digest.h"
+#include "fips.h"
 
 #include "openbsd-compat/openssl-compat.h"
 
@@ -110,8 +111,34 @@ static const struct sshcipher ciphers[]
 	{ NULL,			0, 0, 0, 0, 0, NULL }
 };
 
+static const struct sshcipher ciphers_fips140_2[] = {
+	{ "aes128-cbc",		16, 16, 0, 0, CFLAG_CBC, EVP_aes_128_cbc },
+	{ "aes192-cbc",		16, 24, 0, 0, CFLAG_CBC, EVP_aes_192_cbc },
+	{ "aes256-cbc",		16, 32, 0, 0, CFLAG_CBC, EVP_aes_256_cbc },
+	{ "aes128-ctr",		16, 16, 0, 0, 0, EVP_aes_128_ctr },
+	{ "aes192-ctr",		16, 24, 0, 0, 0, EVP_aes_192_ctr },
+	{ "aes256-ctr",		16, 32, 0, 0, 0, EVP_aes_256_ctr },
+	{ "aes128-gcm@openssh.com",
+				16, 16, 12, 16, 0, EVP_aes_128_gcm },
+	{ "aes256-gcm@openssh.com",
+				16, 32, 12, 16, 0, EVP_aes_256_gcm },
+	{ "none",		8, 0, 0, 0, CFLAG_NONE, NULL },
+
+	{ NULL,			0, 0, 0, 0, 0, NULL }
+};
+
 /*--*/
 
+static const struct sshcipher *
+fips_select_ciphers(void)
+{
+	if (fips_mode()) {
+		return ciphers_fips140_2;
+	}
+
+	return ciphers;
+}
+
 /* Returns a comma-separated list of supported ciphers. */
 char *
 cipher_alg_list(char sep, int auth_only)
@@ -120,7 +147,7 @@ cipher_alg_list(char sep, int auth_only)
 	size_t nlen, rlen = 0;
 	const struct sshcipher *c;
 
-	for (c = ciphers; c->name != NULL; c++) {
+	for (c = fips_select_ciphers(); c->name != NULL; c++) {
 		if ((c->flags & CFLAG_INTERNAL) != 0)
 			continue;
 		if (auth_only && c->auth_len == 0)
@@ -203,7 +230,7 @@ const struct sshcipher *
 cipher_by_name(const char *name)
 {
 	const struct sshcipher *c;
-	for (c = ciphers; c->name != NULL; c++)
+	for (c = fips_select_ciphers(); c->name != NULL; c++)
 		if (strcmp(c->name, name) == 0)
 			return c;
 	return NULL;
Index: os-openssh/fips.c
===================================================================
--- /dev/null
+++ os-openssh/fips.c
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2012 Petr Cerny.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+
+#include "fips.h"
+
+#include "log.h"
+
+int
+fips_mode()
+{
+	static int fips_state = -1;
+	if (-1 == fips_state) {
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+		fips_state = FIPS_mode();
+#else
+		fips_state = EVP_default_properties_is_fips_enabled(NULL);
+#endif
+		if (fips_state)
+			debug("FIPS mode initialized");
+	}
+	return fips_state;
+}
Index: os-openssh/fips.h
===================================================================
--- /dev/null
+++ os-openssh/fips.h
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2012 Petr Cerny.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FIPS_H
+#define FIPS_H
+
+int	 fips_mode(void);
+
+#endif
+
Index: os-openssh/kex-names.c
===================================================================
--- os-openssh.orig/kex-names.c
+++ os-openssh/kex-names.c
@@ -44,6 +44,7 @@
 
 #include "ssherr.h"
 #include "xmalloc.h"
+#include "fips.h"
 
 struct kexalg {
 	char *name;
@@ -90,6 +91,39 @@ static const struct kexalg kexalgs[] = {
 	{ NULL, 0, -1, -1},
 };
 
+static const struct kexalg kexalgs_fips140_2[] = {
+	{ KEX_DH1, KEX_DH_GRP1_SHA1, 0, SSH_DIGEST_SHA1 },
+	{ KEX_DH14_SHA1, KEX_DH_GRP14_SHA1, 0, SSH_DIGEST_SHA1 },
+	{ KEX_DH14_SHA256, KEX_DH_GRP14_SHA256, 0, SSH_DIGEST_SHA256 },
+	{ KEX_DH16_SHA512, KEX_DH_GRP16_SHA512, 0, SSH_DIGEST_SHA512 },
+	{ KEX_DH18_SHA512, KEX_DH_GRP18_SHA512, 0, SSH_DIGEST_SHA512 },
+	{ KEX_DHGEX_SHA1, KEX_DH_GEX_SHA1, 0, SSH_DIGEST_SHA1 },
+#ifdef HAVE_EVP_SHA256
+	{ KEX_DHGEX_SHA256, KEX_DH_GEX_SHA256, 0, SSH_DIGEST_SHA256 },
+#endif
+#ifdef OPENSSL_HAS_ECC
+	{ KEX_ECDH_SHA2_NISTP256, KEX_ECDH_SHA2,
+	    NID_X9_62_prime256v1, SSH_DIGEST_SHA256 },
+	{ KEX_ECDH_SHA2_NISTP384, KEX_ECDH_SHA2, NID_secp384r1,
+	    SSH_DIGEST_SHA384 },
+# ifdef OPENSSL_HAS_NISTP521
+	{ KEX_ECDH_SHA2_NISTP521, KEX_ECDH_SHA2, NID_secp521r1,
+	    SSH_DIGEST_SHA512 },
+# endif
+#endif
+	{ NULL, 0, -1, -1},
+};
+
+static const struct kexalg *
+fips_select_kexalgs(void)
+{
+	if (fips_mode()) {
+		return kexalgs_fips140_2;
+	}
+
+	return kexalgs;
+}
+
 char *
 kex_alg_list(char sep)
 {
@@ -97,7 +131,7 @@ kex_alg_list(char sep)
 	size_t nlen, rlen = 0;
 	const struct kexalg *k;
 
-	for (k = kexalgs; k->name != NULL; k++) {
+	for (k = fips_select_kexalgs(); k->name != NULL; k++) {
 		if (ret != NULL)
 			ret[rlen++] = sep;
 		nlen = strlen(k->name);
@@ -117,7 +151,7 @@ kex_alg_by_name(const char *name)
 {
 	const struct kexalg *k;
 
-	for (k = kexalgs; k->name != NULL; k++) {
+	for (k = fips_select_kexalgs(); k->name != NULL; k++) {
 		if (strcmp(k->name, name) == 0)
 			return k;
 	}
Index: os-openssh/mac.c
===================================================================
--- os-openssh.orig/mac.c
+++ os-openssh/mac.c
@@ -38,6 +38,7 @@
 #include "misc.h"
 #include "ssherr.h"
 #include "sshbuf.h"
+#include "fips.h"
 
 #include "openbsd-compat/openssl-compat.h"
 
@@ -79,6 +80,26 @@ static const struct macalg macs[] = {
 	{ NULL,					0, 0, 0, 0, 0, 0 }
 };
 
+static const struct macalg macs_fips140_2[] = {
+	{ "hmac-sha1",				SSH_DIGEST, SSH_DIGEST_SHA1, 0, 0, 0, 0 },
+	{ "hmac-sha2-256",			SSH_DIGEST, SSH_DIGEST_SHA256, 0, 0, 0, 0 },
+	{ "hmac-sha2-512",			SSH_DIGEST, SSH_DIGEST_SHA512, 0, 0, 0, 0 },
+	{ "hmac-sha1-etm@openssh.com",		SSH_DIGEST, SSH_DIGEST_SHA1, 0, 0, 0, 1 },
+	{ "hmac-sha2-256-etm@openssh.com",	SSH_DIGEST, SSH_DIGEST_SHA256, 0, 0, 0, 1 },
+	{ "hmac-sha2-512-etm@openssh.com",	SSH_DIGEST, SSH_DIGEST_SHA512, 0, 0, 0, 1 },
+	{ NULL,					0, 0, 0, 0, 0, 0 }
+};
+
+static const struct macalg *
+fips_select_macs(void)
+{
+	if (fips_mode()) {
+		return macs_fips140_2;
+	}
+
+	return macs;
+}
+
 /* Returns a list of supported MACs separated by the specified char. */
 char *
 mac_alg_list(char sep)
@@ -87,7 +108,7 @@ mac_alg_list(char sep)
 	size_t nlen, rlen = 0;
 	const struct macalg *m;
 
-	for (m = macs; m->name != NULL; m++) {
+	for (m = fips_select_macs(); m->name != NULL; m++) {
 		if (ret != NULL)
 			ret[rlen++] = sep;
 		nlen = strlen(m->name);
@@ -126,7 +147,7 @@ mac_setup(struct sshmac *mac, char *name
 {
 	const struct macalg *m;
 
-	for (m = macs; m->name != NULL; m++) {
+	for (m = fips_select_macs(); m->name != NULL; m++) {
 		if (strcmp(name, m->name) != 0)
 			continue;
 		if (mac != NULL)
Index: os-openssh/myproposal.h
===================================================================
--- os-openssh.orig/myproposal.h
+++ os-openssh/myproposal.h
@@ -40,6 +40,17 @@
 	"diffie-hellman-group18-sha512," \
 	"diffie-hellman-group14-sha256"
 
+#define KEX_FIPS_140_2_SERVER_KEX \
+	"ecdh-sha2-nistp256," \
+	"ecdh-sha2-nistp384," \
+	"ecdh-sha2-nistp521," \
+	"diffie-hellman-group-exchange-sha256," \
+	"diffie-hellman-group16-sha512," \
+	"diffie-hellman-group18-sha512," \
+	"diffie-hellman-group14-sha256"
+
+#define KEX_FIPS_140_2_CLIENT_KEX KEX_FIPS_140_2_SERVER_KEX
+
 #define	KEX_DEFAULT_PK_ALG	\
 	"ssh-ed25519-cert-v01@openssh.com," \
 	"ecdsa-sha2-nistp256-cert-v01@openssh.com," \
@@ -58,6 +69,20 @@
 	"rsa-sha2-512," \
 	"rsa-sha2-256"
 
+#define	KEX_FIPS_140_2_PK_ALG      \
+	"ecdsa-sha2-nistp256-cert-v01@openssh.com," \
+	"ecdsa-sha2-nistp384-cert-v01@openssh.com," \
+	"ecdsa-sha2-nistp521-cert-v01@openssh.com," \
+	"sk-ecdsa-sha2-nistp256-cert-v01@openssh.com," \
+	"rsa-sha2-512-cert-v01@openssh.com," \
+	"rsa-sha2-256-cert-v01@openssh.com," \
+	"ecdsa-sha2-nistp256," \
+	"ecdsa-sha2-nistp384," \
+	"ecdsa-sha2-nistp521," \
+	"sk-ecdsa-sha2-nistp256@openssh.com," \
+	"rsa-sha2-512," \
+	"rsa-sha2-256"
+
 #define	KEX_SERVER_ENCRYPT \
 	"chacha20-poly1305@openssh.com," \
 	"aes128-gcm@openssh.com,aes256-gcm@openssh.com," \
@@ -65,6 +90,12 @@
 
 #define KEX_CLIENT_ENCRYPT KEX_SERVER_ENCRYPT
 
+#define	KEX_FIPS_140_2_SERVER_ENCRYPT \
+	"aes128-ctr,aes192-ctr,aes256-ctr," \
+	"aes128-gcm@openssh.com,aes256-gcm@openssh.com"
+
+#define	KEX_FIPS_140_2_CLIENT_ENCRYPT KEX_FIPS_140_2_SERVER_ENCRYPT
+
 #define	KEX_SERVER_MAC \
 	"umac-64-etm@openssh.com," \
 	"umac-128-etm@openssh.com," \
@@ -79,6 +110,16 @@
 
 #define KEX_CLIENT_MAC KEX_SERVER_MAC
 
+#define	KEX_FIPS_140_2_SERVER_MAC		\
+	"hmac-sha2-256-etm@openssh.com," \
+	"hmac-sha2-512-etm@openssh.com," \
+	"hmac-sha1-etm@openssh.com," \
+	"hmac-sha2-256," \
+	"hmac-sha2-512," \
+	"hmac-sha1"
+
+#define	KEX_FIPS_140_2_CLIENT_MAC KEX_FIPS_140_2_SERVER_MAC
+
 /* Not a KEX value, but here so all the algorithm defaults are together */
 #define	SSH_ALLOWED_CA_SIGALGS	\
 	"ssh-ed25519," \
@@ -89,6 +130,14 @@
 	"sk-ecdsa-sha2-nistp256@openssh.com," \
 	"rsa-sha2-512," \
 	"rsa-sha2-256"
+
+#define	SSH_FIPS_140_2_ALLOWED_CA_SIGALGS	\
+	"ecdsa-sha2-nistp256," \
+	"ecdsa-sha2-nistp384," \
+	"ecdsa-sha2-nistp521," \
+	"sk-ecdsa-sha2-nistp256@openssh.com," \
+	"rsa-sha2-512," \
+	"rsa-sha2-256"
 
 #define	KEX_DEFAULT_COMP	"none,zlib@openssh.com"
 #define	KEX_DEFAULT_LANG	""
Index: os-openssh/openbsd-compat/arc4random.c
===================================================================
--- os-openssh.orig/openbsd-compat/arc4random.c
+++ os-openssh/openbsd-compat/arc4random.c
@@ -205,13 +205,35 @@ _rs_random_u32(uint32_t *val)
 	rs->rs_have -= sizeof(*val);
 }
 
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+static int
+fips_mode(void)
+{
+	static int fips_state = -1;
+	if (-1 == fips_state) {
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+		fips_state = FIPS_mode();
+#else
+		fips_state = EVP_default_properties_is_fips_enabled(NULL);
+#endif
+		if (fips_state)
+			debug("FIPS mode initialized");
+	}
+	return fips_state;
+}
+
 uint32_t
 arc4random(void)
 {
 	uint32_t val;
 
 	_ARC4_LOCK();
-	_rs_random_u32(&val);
+	if (fips_mode()) {
+		RAND_bytes((uint8_t *)&val, sizeof(val));
+	} else {
+		_rs_random_u32(&val);
+	}
 	_ARC4_UNLOCK();
 	return val;
 }
@@ -226,7 +248,11 @@ void
 arc4random_buf(void *buf, size_t n)
 {
 	_ARC4_LOCK();
-	_rs_random_buf(buf, n);
+	if (fips_mode()) {
+		RAND_bytes(buf, n);
+	} else {
+		_rs_random_buf(buf, n);
+	}
 	_ARC4_UNLOCK();
 }
 DEF_WEAK(arc4random_buf);
Index: os-openssh/readconf.c
===================================================================
--- os-openssh.orig/readconf.c
+++ os-openssh/readconf.c
@@ -71,6 +71,7 @@
 #include "myproposal.h"
 #include "digest.h"
 #include "version.h"
+#include "fips.h"
 
 /* Format of the configuration file:
 
@@ -2989,11 +2990,19 @@ fill_default_options(Options * options)
 	all_key = sshkey_alg_list(0, 0, 1, ',');
 	all_sig = sshkey_alg_list(0, 1, 1, ',');
 	/* remove unsupported algos from default lists */
-	def_cipher = match_filter_allowlist(KEX_CLIENT_ENCRYPT, all_cipher);
-	def_mac = match_filter_allowlist(KEX_CLIENT_MAC, all_mac);
-	def_kex = match_filter_allowlist(KEX_CLIENT_KEX, all_kex);
-	def_key = match_filter_allowlist(KEX_DEFAULT_PK_ALG, all_key);
-	def_sig = match_filter_allowlist(SSH_ALLOWED_CA_SIGALGS, all_sig);
+	if (fips_mode()) {
+		def_cipher = match_filter_allowlist(KEX_FIPS_140_2_CLIENT_ENCRYPT, all_cipher);
+		def_mac = match_filter_allowlist(KEX_FIPS_140_2_CLIENT_MAC, all_mac);
+		def_kex = match_filter_allowlist(KEX_FIPS_140_2_CLIENT_KEX, all_kex);
+		def_key = match_filter_allowlist(KEX_FIPS_140_2_PK_ALG, all_key);
+		def_sig = match_filter_allowlist(SSH_FIPS_140_2_ALLOWED_CA_SIGALGS, all_sig);
+	} else {
+		def_cipher = match_filter_allowlist(KEX_CLIENT_ENCRYPT, all_cipher);
+		def_mac = match_filter_allowlist(KEX_CLIENT_MAC, all_mac);
+		def_kex = match_filter_allowlist(KEX_CLIENT_KEX, all_kex);
+		def_key = match_filter_allowlist(KEX_DEFAULT_PK_ALG, all_key);
+		def_sig = match_filter_allowlist(SSH_ALLOWED_CA_SIGALGS, all_sig);
+	}
 #define ASSEMBLE(what, defaults, all) \
 	do { \
 		if ((r = kex_assemble_names(&options->what, \
Index: os-openssh/servconf.c
===================================================================
--- os-openssh.orig/servconf.c
+++ os-openssh/servconf.c
@@ -69,6 +69,7 @@
 #include "myproposal.h"
 #include "digest.h"
 #include "version.h"
+#include "fips.h"
 
 #if !defined(SSHD_PAM_SERVICE)
 # define SSHD_PAM_SERVICE		"sshd"
@@ -240,11 +241,19 @@ assemble_algorithms(ServerOptions *o)
 	all_key = sshkey_alg_list(0, 0, 1, ',');
 	all_sig = sshkey_alg_list(0, 1, 1, ',');
 	/* remove unsupported algos from default lists */
-	def_cipher = match_filter_allowlist(KEX_SERVER_ENCRYPT, all_cipher);
-	def_mac = match_filter_allowlist(KEX_SERVER_MAC, all_mac);
-	def_kex = match_filter_allowlist(KEX_SERVER_KEX, all_kex);
-	def_key = match_filter_allowlist(KEX_DEFAULT_PK_ALG, all_key);
-	def_sig = match_filter_allowlist(SSH_ALLOWED_CA_SIGALGS, all_sig);
+	if (fips_mode()) {
+		def_cipher = match_filter_allowlist(KEX_FIPS_140_2_SERVER_ENCRYPT, all_cipher);
+		def_mac = match_filter_allowlist(KEX_FIPS_140_2_SERVER_MAC, all_mac);
+		def_kex = match_filter_allowlist(KEX_FIPS_140_2_SERVER_KEX, all_kex);
+		def_key = match_filter_allowlist(KEX_FIPS_140_2_PK_ALG, all_key);
+		def_sig = match_filter_allowlist(SSH_FIPS_140_2_ALLOWED_CA_SIGALGS, all_sig);
+	} else {
+		def_cipher = match_filter_allowlist(KEX_SERVER_ENCRYPT, all_cipher);
+		def_mac = match_filter_allowlist(KEX_SERVER_MAC, all_mac);
+		def_kex = match_filter_allowlist(KEX_SERVER_KEX, all_kex);
+		def_key = match_filter_allowlist(KEX_DEFAULT_PK_ALG, all_key);
+		def_sig = match_filter_allowlist(SSH_ALLOWED_CA_SIGALGS, all_sig);
+	}
 #define ASSEMBLE(what, defaults, all) \
 	do { \
 		if ((r = kex_assemble_names(&o->what, defaults, all)) != 0) \
@@ -2650,6 +2659,9 @@ process_server_config_line_depth(ServerO
 		if ((value = ssh_digest_alg_by_name(arg)) == -1)
 			fatal("%.200s line %d: Invalid %s algorithm \"%s\".",
 			    filename, linenum, keyword, arg);
+		if (fips_mode() && value == SSH_DIGEST_MD5)
+			fatal("%.200s line %d: %s algorithm not permitted \"%s\".",
+			    filename, linenum, keyword, arg);
 		if (*activep)
 			options->fingerprint_hash = value;
 		break;
Index: os-openssh/sshkey.c
===================================================================
--- os-openssh.orig/sshkey.c
+++ os-openssh/sshkey.c
@@ -60,6 +60,7 @@
 #include "sshkey.h"
 #include "match.h"
 #include "ssh-sk.h"
+#include "fips.h"
 
 #ifdef WITH_XMSS
 #include "sshkey-xmss.h"
@@ -170,14 +171,56 @@ const struct sshkey_impl * const keyimpl
 	NULL
 };
 
+static const struct sshkey_impl * const keyimpls_fips140_2[] = {
+#ifdef WITH_OPENSSL
+# ifdef OPENSSL_HAS_ECC
+	&sshkey_ecdsa_nistp256_impl,
+	&sshkey_ecdsa_nistp256_cert_impl,
+	&sshkey_ecdsa_nistp384_impl,
+	&sshkey_ecdsa_nistp384_cert_impl,
+#  ifdef OPENSSL_HAS_NISTP521
+	&sshkey_ecdsa_nistp521_impl,
+	&sshkey_ecdsa_nistp521_cert_impl,
+#  endif /* OPENSSL_HAS_NISTP521 */
+#  ifdef ENABLE_SK
+	&sshkey_ecdsa_sk_impl,
+	&sshkey_ecdsa_sk_cert_impl,
+	&sshkey_ecdsa_sk_webauthn_impl,
+#  endif /* ENABLE_SK */
+# endif /* OPENSSL_HAS_ECC */
+# ifdef WITH_DSA
+	&sshkey_dss_impl,
+	&sshkey_dsa_cert_impl,
+# endif
+	&sshkey_rsa_impl,
+	&sshkey_rsa_cert_impl,
+	&sshkey_rsa_sha256_impl,
+	&sshkey_rsa_sha256_cert_impl,
+	&sshkey_rsa_sha512_impl,
+	&sshkey_rsa_sha512_cert_impl,
+#endif /* WITH_OPENSSL */
+	NULL
+};
+
+static const struct sshkey_impl * const *
+fips_select_keyimpls(void)
+{
+	if (fips_mode()) {
+		return keyimpls_fips140_2;
+	}
+
+	return keyimpls;
+}
+
 static const struct sshkey_impl *
 sshkey_impl_from_type(int type)
 {
 	int i;
+	const struct sshkey_impl * const *impls = fips_select_keyimpls();
 
-	for (i = 0; keyimpls[i] != NULL; i++) {
-		if (keyimpls[i]->type == type)
-			return keyimpls[i];
+	for (i = 0; impls[i] != NULL; i++) {
+		if (impls[i]->type == type)
+			return impls[i];
 	}
 	return NULL;
 }
@@ -186,11 +229,12 @@ static const struct sshkey_impl *
 sshkey_impl_from_type_nid(int type, int nid)
 {
 	int i;
+	const struct sshkey_impl * const *impls = fips_select_keyimpls();
 
-	for (i = 0; keyimpls[i] != NULL; i++) {
-		if (keyimpls[i]->type == type &&
-		    (keyimpls[i]->nid == 0 || keyimpls[i]->nid == nid))
-			return keyimpls[i];
+	for (i = 0; impls[i] != NULL; i++) {
+		if (impls[i]->type == type &&
+		    (impls[i]->nid == 0 || impls[i]->nid == nid))
+			return impls[i];
 	}
 	return NULL;
 }
@@ -251,9 +295,10 @@ type_from_name(const char *name, int all
 {
 	int i;
 	const struct sshkey_impl *impl;
+	const struct sshkey_impl * const *impls = fips_select_keyimpls();
 
-	for (i = 0; keyimpls[i] != NULL; i++) {
-		impl = keyimpls[i];
+	for (i = 0; impls[i] != NULL; i++) {
+		impl = impls[i];
 		if (impl->name != NULL && strcmp(name, impl->name) == 0)
 			return impl->type;
 		/* Only allow shortname matches for plain key types */
@@ -293,13 +338,14 @@ int
 sshkey_ecdsa_nid_from_name(const char *name)
 {
 	int i;
+	const struct sshkey_impl * const *impls = fips_select_keyimpls();
 
-	for (i = 0; keyimpls[i] != NULL; i++) {
-		if (!key_type_is_ecdsa_variant(keyimpls[i]->type))
+	for (i = 0; impls[i] != NULL; i++) {
+		if (!key_type_is_ecdsa_variant(impls[i]->type))
 			continue;
-		if (keyimpls[i]->name != NULL &&
-		    strcmp(name, keyimpls[i]->name) == 0)
-			return keyimpls[i]->nid;
+		if (impls[i]->name != NULL &&
+		    strcmp(name, impls[i]->name) == 0)
+			return impls[i]->nid;
 	}
 	return -1;
 }
@@ -333,9 +379,10 @@ sshkey_alg_list(int certs_only, int plai
 	char *tmp, *ret = NULL;
 	size_t i, nlen, rlen = 0;
 	const struct sshkey_impl *impl;
+	const struct sshkey_impl * const *impls = fips_select_keyimpls();
 
-	for (i = 0; keyimpls[i] != NULL; i++) {
-		impl = keyimpls[i];
+	for (i = 0; impls[i] != NULL; i++) {
+		impl = impls[i];
 		if (impl->name == NULL)
 			continue;
 		if (!include_sigonly && impl->sigonly)
@@ -377,11 +424,12 @@ sshkey_names_valid2(const char *names, i
 				 * If any has a positive or negative match then
 				 * the component is accepted.
 				 */
+				const struct sshkey_impl * const *impls = fips_select_keyimpls();
 				impl = NULL;
-				for (i = 0; keyimpls[i] != NULL; i++) {
+				for (i = 0; impls[i] != NULL; i++) {
 					if (match_pattern_list(
-					    keyimpls[i]->name, p, 0) != 0) {
-						impl = keyimpls[i];
+					    impls[i]->name, p, 0) != 0) {
+						impl = impls[i];
 						break;
 					}
 				}
@@ -1291,9 +1339,10 @@ peek_type_nid(const char *s, size_t l, i
 {
 	const struct sshkey_impl *impl;
 	int i;
+	const struct sshkey_impl * const *impls = fips_select_keyimpls();
 
-	for (i = 0; keyimpls[i] != NULL; i++) {
-		impl = keyimpls[i];
+	for (i = 0; impls[i] != NULL; i++) {
+		impl = impls[i];
 		if (impl->name == NULL || strlen(impl->name) != l)
 			continue;
 		if (memcmp(s, impl->name, l) == 0) {
@@ -2161,9 +2210,10 @@ sshkey_sigalg_by_name(const char *name)
 {
 	const struct sshkey_impl *impl;
 	int i;
+	const struct sshkey_impl * const *impls = fips_select_keyimpls();
 
-	for (i = 0; keyimpls[i] != NULL; i++) {
-		impl = keyimpls[i];
+	for (i = 0; impls[i] != NULL; i++) {
+		impl = impls[i];
 		if (strcmp(impl->name, name) != 0)
 			continue;
 		if (impl->sigalg != NULL)
@@ -3450,6 +3500,10 @@ sshkey_private_to_fileblob(struct sshkey
 #ifdef WITH_OPENSSL
 	case KEY_ECDSA_SK:
 #endif /* WITH_OPENSSL */
+		if (fips_mode() && key->type != KEY_ECDSA_SK) {
+			/* ED25519 not supported in FIPS mode */
+			return SSH_ERR_KEY_TYPE_UNKNOWN;
+		}
 		return sshkey_private_to_blob2(key, blob, passphrase,
 		    comment, openssh_format_cipher, openssh_format_rounds);
 	default:
@@ -3457,6 +3511,8 @@ sshkey_private_to_fileblob(struct sshkey
 	}
 
 #ifdef WITH_OPENSSL
+	if (fips_mode())
+	    format = SSHKEY_PRIVATE_PEM;
 	switch (format) {
 	case SSHKEY_PRIVATE_OPENSSH:
 		return sshkey_private_to_blob2(key, blob, passphrase,
@@ -3737,13 +3793,22 @@ sshkey_parse_private_fileblob_type(struc
 		*commentp = NULL;
 
 	switch (type) {
+ 	case KEY_ED25519:
 	case KEY_XMSS:
+		if (fips_mode()) {
+			/* ED25519 keys unavailable in FIPS mode */
+			return SSH_ERR_INVALID_FORMAT;
+		}
 		/* No fallback for new-format-only keys */
 		return sshkey_parse_private2(blob, type, passphrase,
 		    keyp, commentp);
 	default:
-		r = sshkey_parse_private2(blob, type, passphrase, keyp,
-		    commentp);
+		if (!fips_mode()) {
+			r = sshkey_parse_private2(blob, type, passphrase, keyp,
+			    commentp);
+		} else {
+			r = SSH_ERR_INVALID_FORMAT; /* to force fallback below */
+		}
 		/* Only fallback to PEM parser if a format error occurred. */
 		if (r != SSH_ERR_INVALID_FORMAT)
 			return r;
